File: 프로그래머스 level 2/주차 요금 계산.py
from collections import defaultdict
import math

def solution(fees, records):
    
    fee_record_dict = defaultdict(list)
    
    fee_list = []
    
    for record in records:
        record_time, car_num, in_or_out = record.split()
        fee_record_dict[car_num].append(time_to_min(record_time))
    
    for key, value in fee_record_dict.items():
        total_fee = calculate_fee(value, fees)
        fee_list.append((key, total_fee))
        
    fee_list.sort(key = lambda x:x[0])
    
    return [item[1] for item in fee_list]
        
def time_to_min(string_time):
    hour, minute = string_time.split(":")
    total_minute = int(hour) * 60 + int(minute)
    
    return total_minute

def calculate_fee(time_list, fees) -> int:
    BASIC_TIME, BASIC_FEE, OVER_TIME, OVER_FEE = fees
    LAST_TIME = time_to_min("23:59")
    
    total_fee = 0
    total_parking_time = 0
    
    if len(time_list) % 2 == 1:
        time_list.append(LAST_TIME)
        
    for time_index in range(len(time_list) // 2):
        in_time, out_time = time_list[time_index * 2], time_list[time_index * 2 + 1]
        
        total_parking_time += out_time - in_time
    
    if total_parking_time <= BASIC_TIME:
        total_fee += BASIC_FEE
    else:
        total_fee += BASIC_FEE + math.ceil((total_parking_time - BASIC_TIME) / OVER_TIME) * OVER_FEE

    return total_fee

File: Swift/프로그래머스 level 2/주차요금계산.swift
import Foundation

func solution(_ fees:[Int], _ records:[String]) -> [Int] {
    var inOutRecordDict: [String: [Int]] = [:]
    
    var answerList: [(String, Int)] = []
    
    let time = 0
    let carNum = 1
    
    records.forEach {
        let inOutData = $0.components(separatedBy: " ")
        
        let inOutTime = stringToMinute(inOutData[time])
        let inOutCarNum = inOutData[carNum]
        inOutRecordDict[inOutCarNum, default: []].append(inOutTime)
    }
    
    inOutRecordDict.forEach {
        let totalFee = calculateFee($1, fees)
        
        answerList.append(($0, totalFee))
    }
    
    answerList.sort(by: {$0.0 < $1.0})
    
    return answerList.map { $0.1 }
}

func stringToMinute(_ inOutTime: String) -> Int {
    let timeData = inOutTime.components(separatedBy: ":")
    
    let hour = Int(timeData[0])!
    let minute = Int(timeData[1])!
    
    let totalMinute = hour * 60 + minute
    
    return totalMinute
}

func calculateFee(_ timeList: [Int], _ feeList: [Int]) -> Int {
    var totalFee: Int = 0
    
    var recordedTimeList = timeList
    let latestTime = 23 * 60 + 59
    
    if recordedTimeList.count % 2 == 1 {
        recordedTimeList.append(latestTime)
    }
    
    var totalTime = 0

    for idx in 0..<Int(recordedTimeList.count / 2) {
        totalTime += recordedTimeList[idx * 2 + 1] - recordedTimeList[idx * 2]
    }
    
    let basicTime = 0
    let basicFee = 1
    let additionalTime = 2
    let additionalFee = 3

    if totalTime <= feeList[basicTime] {
        totalFee = feeList[basicFee]
    } else {
        let overTimeFee = Int(ceil(Double(totalTime - feeList[basicTime]) / Double(feeList[additionalTime]))) * feeList[additionalFee]
        
        totalFee = feeList[basicFee] + overTimeFee
    }
    
    return totalFee
}